lib/smack/SmackRep.cpp

```diff
@@ -758,36 +938,96 @@ const Expr* SmackRep::declareIsExternal(const Expr* e) {
 }
 
 string SmackRep::getPtrType() {
-  return "int";
+    return "ref";
 }
 
 string SmackRep::memcpyProc(int dstReg, int srcReg) {
   stringstream s;
 
   if (SmackOptions::MemoryModelImpls) {
-    s << "procedure $memcpy." << dstReg << "." << srcReg;
-    s << "(dest: int, src: int, len: int, align: int, isvolatile: bool)" << endl;
-    s << "modifies " << memReg(dstReg) << ";" << endl;
-    s << "{" << endl;
-    s << "  var $oldSrc: [" << getPtrType() << "] " << getPtrType() << ";" << endl;
-    s << "  var $oldDst: [" << getPtrType() << "] " << getPtrType() << ";" << endl;
-    s << "  $oldSrc := " << memReg(srcReg) << ";" << endl;
-    s << "  $oldDst := " << memReg(dstReg) << ";" << endl;
-    s << "  havoc " << memReg(dstReg) << ";" << endl;
-    s << "  assume (forall x:int :: dest <= x && x < dest + len ==> "
-      << memReg(dstReg) << "[x] == $oldSrc[src - dest + x]);" << endl;
-    s << "  assume (forall x:int :: !(dest <= x && x < dest + len) ==> "
-      << memReg(dstReg) << "[x] == $oldDst[x]);" << endl;
-    s << "}" << endl;
+    if (SmackOptions::BitVectors) {
```

Comment:
Could these two branches be merged?

First, the `memPath` function should be use in both — the second branch is using `memReg` instead.

Second, couldn't the variables and assignments in loops be rewritten using the existing $load and $store style bit concatenation?